### PR TITLE
Spark 3.3: Add Default Parallelism Level for All Spark Driver Based Deletes

### PR DESCRIPTION
**Pull request by RussellSpitzer**
An issue we've run into frequently is that several Spark actions perform deletes on the driver with a default parallelism of 1. This is quite slow for S3 and painfully slow for very large tables. To fix this we change the default behavior to always be multithreaded deletes.

The default for all Spark related actions can then be changed with a SQL Conf parameter as well as within each command with their own parallelism parameters.

### CODE REVIEW

**Comment by RussellSpitzer** [comment]
@anuragmantri + @aokolnychyi + @rdblue - This is a bit of a big default behavior change but it's been biting a lot of our users lately and the change is relatively safe.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java [hunk 1]
```diff
@@ -47,4 +47,8 @@ private SparkSQLProperties() {}
   public static final String PRESERVE_DATA_GROUPING =
       "spark.sql.iceberg.planning.preserve-data-grouping";
   public static final boolean PRESERVE_DATA_GROUPING_DEFAULT = false;
+
+  // Controls how many physical file deletes to execute in parallel when not otherwise specified
+  public static final String DELETE_PARALLELISM = "driver-delete-default-parallelism";
+  public static final String DELETE_PARALLELISM_DEFAULT = "25";
```
Comment (RussellSpitzer) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:53]: With S3's request throttling around 4k requests a second this gives us a lot of overhead. 
Assuming a 50ms response time
4000 max requests / Second / 20 requests per thread per second =~  200 max concurrent requests. 

Another option for this is to also incorporate the "bulk delete" apis but that would only help with S3 based filesystems.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java [hunk 2]
```diff
@@ -231,24 +258,27 @@ protected DeleteSummary deleteFiles(
...
-            })
-        .run(
-            fileInfo -> {
-              String path = fileInfo.getPath();
-              String type = fileInfo.getType();
-              deleteFunc.accept(path);
-              summary.deletedFile(path, type);
-            });
+
+    withDefaultDeleteService(executorService, (deleteService) -> {
```
Comment (RussellSpitzer) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java:262]: This covers ExpireSnapshots and DropTable with Purge

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java [hunk 3]
```diff
@@ -246,12 +246,13 @@ private DeleteOrphanFiles.Result doExecute() {
     List<String> orphanFiles =
         findOrphanFiles(spark(), actualFileIdentDS, validFileIdentDS, prefixMismatchMode);
 
-    Tasks.foreach(orphanFiles)
-        .noRetry()
-        .executeWith(deleteExecutorService)
-        .suppressFailureWhenFinished()
-        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
-        .run(deleteFunc::accept);
+    withDefaultDeleteService(deleteExecutorService, deleteService ->
```
Comment (RussellSpitzer) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:249]: This change covers RemoveOrphanFiles

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java [hunk 4]
```diff
@@ -47,4 +47,8 @@ private SparkSQLProperties() {}
   public static final String PRESERVE_DATA_GROUPING =
       "spark.sql.iceberg.planning.preserve-data-grouping";
   public static final boolean PRESERVE_DATA_GROUPING_DEFAULT = false;
+
+  // Controls how many physical file deletes to execute in parallel when not otherwise specified
+  public static final String DELETE_PARALLELISM = "driver-delete-default-parallelism";
```
Comment (anuragmantri) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:52]: Should we change this to `spark.sql.iceberg.driver-delete-default-parallelism` to be consistent with the rest of the properties in this file?

**Review by anuragmantri — APPROVED** [review]
Thanks for this change @RussellSpitzer. It looks safe to me.

I left a minor comment and build failed with code format violations, you may want to run `gradlew spotlessApply` on the file.

**Comment by aokolnychyi** [comment]
Let me take a look soon.

**Comment by RussellSpitzer** [comment]
Talking with @aokolnychyi we decided we are going to take a slightly broader approach here. Rather than allowing each action to define it's own method of deleting and using custom executor services, we will fall back to each FileIO's bulk delete support. We will then add a basic parallel default delete to HDFS for bulk delete.

After doing this we will deprecate all the "parallel delete" methods from the actions and procedures instead instructing users to use their IO specific parallelism controls.

**Comment by amogh-jahagirdar** [comment]
Thanks for clarifying @RussellSpitzer I think it makes a ton of sense to leave the specifics of bulk vs parallel to the FileIO abstraction. In this case, we leverage bulk delete wherever possible (S3) and then do parallel deletions for file systems which don't support bulk ops like HDFS to improve the throughput of deletions

**Comment by aokolnychyi** [comment]
cc #5373 #5375

**Pull request closed without merging**